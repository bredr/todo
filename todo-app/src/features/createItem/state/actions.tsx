import { ActionType, createAction } from "typesafe-actions";

export const open = createAction("OPEN_CREATE");
export const close = createAction("CLOSE_CREATE");

export type Actions = ActionType<typeof open> | ActionType<typeof close>;

export default { open, close };
