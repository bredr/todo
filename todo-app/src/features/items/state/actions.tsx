import { ActionType, createAction } from "typesafe-actions";
import { grpcRequest } from "../../../middleware/grpc";
import { Items, Pagination } from "../../../proto/todo_pb";

export const setItems = createAction("SET_ITEMS", action => {
  return (items: Items.AsObject) => action(items);
});

export const loadingItems = createAction("LOADING_ITEMS");

export const getItems = grpcRequest<Pagination, Items>();

export type Actions =
  | ActionType<typeof setItems>
  | ActionType<typeof loadingItems>
  | ActionType<typeof getItems>;
