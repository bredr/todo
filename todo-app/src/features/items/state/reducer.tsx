import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Item } from "../../../proto/todo_pb";
import {
  Actions,
  cancelEditItem,
  editItem,
  getItems,
  loadingItems,
  setItems
} from "./actions";

interface IEditableItem extends Item.AsObject {
  edit: boolean;
}

export interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: IEditableItem[];
  readonly loading: boolean;
}

const initState: IState = {
  items: [],
  limit: 100,
  loading: false,
  offset: 0,
  total: 0
};

export default createReducer<IState, Actions>(initState)
  .handleAction(loadingItems, (state, action) =>
    produce(state, draft => {
      draft.loading = true;
      draft.items = [];
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
    })
  )
  .handleAction(setItems, (state, action) =>
    produce(state, draft => {
      draft.loading = false;
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
      draft.total = action.payload.total;
      draft.items = action.payload.itemsList.map(item => ({
        ...item,
        edit: false
      }));
    })
  )
  .handleAction(editItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: true } : item
      );
    })
  )
  .handleAction(cancelEditItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: false } : item
      );
    })
  );
