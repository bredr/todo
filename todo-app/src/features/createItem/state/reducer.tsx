import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, close, open } from "./actions";

export interface IState {
  readonly isOpen: boolean;
}

const initState: IState = {
  isOpen: false
};

export default createReducer<IState, Actions>(initState)
  .handleAction(open, state =>
    produce(state, draft => {
      draft.isOpen = true;
    })
  )
  .handleAction(close, state =>
    produce(state, draft => {
      draft.isOpen = false;
    })
  );
