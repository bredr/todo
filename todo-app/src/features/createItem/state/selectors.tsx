import { createSelector } from "reselect";
import { IState } from "./reducer";

export const CREATE_SLICE = "CREATE_SLICE";

const sliceSelector = (state: { [CREATE_SLICE]: IState }) =>
  state[CREATE_SLICE];

const createItemOpen = createSelector(
  sliceSelector,
  state => state.isOpen
);

export default { createItemOpen };
