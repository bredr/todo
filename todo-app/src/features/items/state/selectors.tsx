import { createSelector } from "reselect";
import { IState } from "./reducer";

export const ITEMS_SLICE = "ITEMS_SLICE";

const sliceSelector = (state: { [ITEMS_SLICE]: IState }) => state[ITEMS_SLICE];

const items = createSelector(
  sliceSelector,
  state => state.items
);
const loading = createSelector(
  sliceSelector,
  state => state.loading
);
const hasNext = createSelector(
  sliceSelector,
  state => state.total > 0 && state.total - (state.limit + state.offset) > 0
);
const hasPrevious = createSelector(
  sliceSelector,
  state => state.offset > 0 && state.total > 0
);

export default { items, loading, hasNext, hasPrevious };
