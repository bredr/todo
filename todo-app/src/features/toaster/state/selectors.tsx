import { createSelector } from "reselect";
import { IState } from "./reducer";

export const TOASTER_SLICE = "TOASTER_SLICE";

const sliceSelector = (state: { [TOASTER_SLICE]: IState }) =>
  state[TOASTER_SLICE];

const currentMessage = createSelector(
  sliceSelector,
  state => (state.messages.length > 0 ? state.messages[0] : undefined)
);

export default { currentMessage };
