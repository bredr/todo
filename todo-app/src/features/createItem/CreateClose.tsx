import { Button } from "@material-ui/core";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { cancelEditItem } from "../items/state/actions";
import { close } from "./state/actions";
import selectors from "./state/selectors";

interface IProps {
  id?: string;
}
export const CreateClose: React.FC<IProps> = props => {
  const { id } = props;
  const dispatch = useDispatch();

  const isOpen = useSelector(selectors.createItemOpen);

  function onClick() {
    id ? dispatch(cancelEditItem(id)) : dispatch(close());
  }

  return (
    <Button
      variant="outlined"
      color="primary"
      fullWidth={true}
      disabled={!isOpen && !id}
      onClick={onClick}
    >
      Cancel
    </Button>
  );
};
