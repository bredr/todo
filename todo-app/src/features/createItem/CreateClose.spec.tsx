import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { CreateClose } from "./CreateClose";
import { CREATE_SLICE } from "./state/selectors";

const mockStore = configureStore([]);

describe("CreateClose", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const store = mockStore({ [CREATE_SLICE]: { isOpen: true } });
    const { container } = render(
      <Provider store={store}>
        <CreateClose />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("dispatches close if no id", () => {
    const store = mockStore({ [CREATE_SLICE]: { isOpen: true } });
    const { getByLabelText } = render(
      <Provider store={store}>
        <CreateClose />
      </Provider>
    );
    fireEvent.click(getByLabelText("close"));
    expect(store.getActions()[0]).toEqual(
      expect.objectContaining({ type: "CLOSE_CREATE" })
    );
  });
  it("dispatches cancelEditItem if no id", () => {
    const store = mockStore({ [CREATE_SLICE]: { isOpen: true } });
    const props = { id: "id" };
    const { getByLabelText } = render(
      <Provider store={store}>
        <CreateClose {...props} />
      </Provider>
    );
    fireEvent.click(getByLabelText("close"));
    expect(store.getActions()[0]).toEqual(
      expect.objectContaining({ payload: "id", type: "CANCEL_EDIT_ITEM" })
    );
  });
});
