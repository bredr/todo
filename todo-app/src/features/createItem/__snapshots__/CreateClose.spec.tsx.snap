// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateClose renders 1`] = `
<button
  aria-label="close"
  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Cancel
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
